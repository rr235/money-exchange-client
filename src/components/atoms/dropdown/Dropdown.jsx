import React, { useState } from 'react';
import classNames from 'classnames';
import styles from './dropdown.styles.scss';

const getSelection = (options, selectedValue) => {
  const selection = options.filter((option) => option.value === selectedValue);
  return selection.length ? selection[0].text : options[0].text;
};

const DropdownListItem = ({ value, text, selected, onClick }) => {
  const onClickHandler = () => {
    onClick({ value, text });
  };

  return (
    <li
      data-value={value}
      role="option"
      aria-selected={selected}
      key={value}
      onClick={onClickHandler}
      className={styles.option}
    >
      {text}
    </li>
  );
};

const Dropdown = ({ options, onSelect, className, selectedValue }) => {
  const [showOptions, setShowOptions] = useState(false);

  const optionsClassName = classNames(styles.options, {
    [styles.show]: showOptions,
  });
  const dropdownClassName = classNames(styles.dropdown, className);

  const onOptionSelectHandler = ({ text, value }) => {
    setShowOptions(false);
    onSelect({ text, value });
  };

  const onClickHandler = () => {
    // toggle options
    setShowOptions(!showOptions);
  };

  return (
    <div className={dropdownClassName}>
      <button
        type="button"
        aria-haspopup="listbox"
        onClick={onClickHandler}
        className={styles.button}
      >
        {getSelection(options, selectedValue)}
      </button>
      <ul role="listbox" className={optionsClassName}>
        {options.map(({ value, text, selected }) => (
          <DropdownListItem
            value={value}
            text={text}
            selected={selected}
            key={value}
            onClick={onOptionSelectHandler}
          />
        ))}
      </ul>
    </div>
  );
};

export default Dropdown;
