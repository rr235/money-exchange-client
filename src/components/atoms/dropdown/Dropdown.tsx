import React, { useState } from 'react';
import classNames from 'classnames';
import styles from './dropdown.styles.scss';

type Option = {
  value: number | string;
  text: string;
  selected?: boolean;
  onClick?(value: Option): void;
};

type props = {
  className?: string;
  options: Option[];
  onSelect?(value: Option): void;
};

const getSelection = (options: Option[]): string => {
  const selection = options.filter((option) => option.selected);
  return selection.length ? selection[0].text : options[0].text;
};

const DropdownListItem = ({ value, text, selected, onClick }: Option) => {
  const onClickHandler = () => {
    onClick({ value, text });
  };

  return (
    <li
      data-value={value}
      role="option"
      aria-selected={selected}
      key={value}
      onClick={onClickHandler}
    >
      {text}
    </li>
  );
};

const Dropdown = ({ options, onSelect, className }: props) => {
  const [showOptions, setShowOptions] = useState<boolean>(false);

  const optionsClassName = classNames(styles.options, {
    [styles.show]: showOptions,
  });
  const buttonClassName = classNames(styles.button, className);

  const onOptionSelectHandler = ({ text, value }: Option) => {
    onSelect({ text, value });
  };

  const onClickHandler = () => {
    // toggle options
    setShowOptions(!showOptions);
  };

  return (
    <>
      <button
        type="button"
        aria-haspopup="listbox"
        onClick={onClickHandler}
        className={buttonClassName}
      >
        {getSelection(options)}
      </button>
      <ul role="listbox" className={optionsClassName}>
        {options.map(({ value, text, selected }: Option) => (
          <DropdownListItem
            value={value}
            text={text}
            selected={selected}
            key={value}
            onClick={onOptionSelectHandler}
          />
        ))}
      </ul>
    </>
  );
};

export default Dropdown;
