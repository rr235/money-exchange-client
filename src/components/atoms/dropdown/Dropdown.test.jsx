import React from 'react';
import renderer from 'react-test-renderer';
import { shallow, mount } from 'enzyme';
import { act } from 'react-dom/test-utils';
import Dropdown, { DropdownListItem } from './Dropdown';

const options = [
  { text: 'option1', value: 'val1' },
  { text: 'option2', value: 'val2' },
  { text: 'option3', value: 'val3' },
];

describe('Dropdown', () => {
  let component;
  beforeAll(() => {
    component = shallow(<Dropdown options={options} id="id" />);
  });

  it('should render correctly', () => {
    const tree = renderer
      .create(<Dropdown options={options} id="id" selectedValue="val2" />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should select first option by default', () => {
    const button = component.find('button');
    expect(button.text()).toBe('option1');
  });

  it('should set correct number of options', () => {
    const dropdownOptions = component.find(DropdownListItem);
    expect(dropdownOptions).toHaveLength(options.length);
  });

  it('options should be hidden by default', () => {
    const dropdownOptions = component.find('ul[role="listbox"]');
    const classList = dropdownOptions.prop('className');
    expect(classList.indexOf('show')).toBe(-1);
  });
});

describe('Dropdown (functionality)', () => {
  let component;
  const mockSelectCallBack = jest.fn();
  beforeEach(() => {
    act(() => {
      component = mount(
        <Dropdown options={options} id="id" onSelect={mockSelectCallBack} />
      );
    });
  });

  it('should show options on activation', () => {
    const button = component.find('button');
    let dropdownOptions = component.find('ul[role="listbox"]');
    let classList = dropdownOptions.prop('className');
    expect(classList.indexOf('show')).toBe(-1); // confirms options are hidden

    // simulate dropdown activation
    act(() => {
      button.props().onClick();
    });

    // update state and get latest ref to options
    component.update();
    dropdownOptions = component.find('ul[role="listbox"]');
    classList = dropdownOptions.prop('className');

    expect(classList.indexOf('show')).toBeGreaterThan(-1);
  });

  it('should select correct options on activation', () => {
    const secondOption = component.find('li').at(1);
    const button = component.find('button');
    expect(button.text()).toBe('option1');

    // simulate selection of second option
    act(() => {
      secondOption.props().onClick();
    });

    expect(mockSelectCallBack).toHaveBeenCalledTimes(1);
    expect(mockSelectCallBack).toHaveBeenCalledWith({
      text: 'option2',
      value: 'val2',
    });
  });
});
