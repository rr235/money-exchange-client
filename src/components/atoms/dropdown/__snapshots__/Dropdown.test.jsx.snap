// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown should render correctly 1`] = `
<div
  className="dropdown"
>
  <button
    aria-haspopup="listbox"
    className="button"
    id="id"
    onClick={[Function]}
    type="button"
  >
    option2
  </button>
  <ul
    className="options"
    role="listbox"
  >
    <li
      aria-selected={false}
      className="option"
      data-value="val1"
      onClick={[Function]}
      onKeyDown={[Function]}
      role="option"
      tabIndex="0"
    >
      option1
    </li>
    <li
      aria-selected={false}
      className="option"
      data-value="val2"
      onClick={[Function]}
      onKeyDown={[Function]}
      role="option"
      tabIndex="0"
    >
      option2
    </li>
    <li
      aria-selected={false}
      className="option"
      data-value="val3"
      onClick={[Function]}
      onKeyDown={[Function]}
      role="option"
      tabIndex="0"
    >
      option3
    </li>
  </ul>
</div>
`;
